✨ Feature: #15 동적 조이스틱 및 조작 UX 개선

Gojaehyeon opened the pull request:
<!--
🙏 PR 제목 컨벤션 (Gitmoji + 타입 + 이슈 번호 + 작업 요약)
예시: ✨ Feat: #167 예약 취소 구현
※ PR 생성 시 Assignees 및 Labels 설정도 잊지 마세요!
-->

## ✨ What’s this PR?
### 📌 관련 이슈 (Related Issue)
<!-- 해당 PR이 어떤 이슈를 해결하는지 연결해주세요 -->
- Closes #15
- 동적 조이스틱 및 조작 UX 개선
---

### 🧶 주요 변경 내용 (Summary)
<!-- 이번 PR에서 작업한 핵심 변경 사항을 작성해주세요 -->

- 동적 조이스틱 시스템 구현: 화면 아무 곳이나 터치 시 해당 위치에 조이스틱 생성, 손가락을 떼면 조이스틱이 기본 위치로 복귀(기존 조이스틱은 터치 시마다 새로 생성/제거됨)
- 조이스틱 투명도 조정: 베이스(alpha) 0.2, 노브(alpha) 0.4로 시인성 개선
- 조이스틱 위치 조정
기본 위치를 기존보다 왼쪽으로 30px 이동 (-230)
- 조이스틱 세기에 따른 속도 조절: 조이스틱을 많이 당길수록 플레이어 속도가 더 빨라짐, 최대 속도(maxSpeed)를 400으로 상향
- 컴퓨터와의 채팅 상황에서 조이스틱 생성되지 않도록 조치

---

### 📸 스크린샷 (Optional)


https://github.com/user-attachments/assets/fcc22208-4179-439a-b494-a9a37dd066c5


---

### 🧪 테스트 / 검증 내역
<!-- 동작 확인 여부나 시나리오 테스트 내용을 간단히 써주세요 -->

- [x] UI 정상 동작 확인
- [x] iPhone 15, iOS 17.4 환경에서 정상 동작
- [ ] 다크모드 테스트는 이후 이슈로 분리 예정 (#000)

---

### 💬 기타 공유 사항
<!-- 리뷰어가 참고하면 좋을 정보, 고민했던 지점 등을 적어주세요 -->


---

### 🙇🏻‍♀️ 리뷰 가이드 (선택)
<!-- 리뷰어가 중점적으로 보면 좋을 포인트가 있다면 알려주세요 -->

Gojaehyeon committed:
feat: joystick 투명도 조정

Gojaehyeon committed:
feat: joystick 기본 위치 이동

Gojaehyeon committed:
feat: 다른 화면 터치시 조이스틱 생성 기능 구현

- 조이스틱 생성시 기존 조이스틱 제거
- 터치 해제시 기존 조이스틱 생성
- 조이스틱 사용 방식은 기존과 동일

Gojaehyeon committed:
feat: 조이스틱을 당기는 만큼 속도가 빨라지는 기능 구현

- 최고속도는 400

yangsijun reviewed (dismissed):
<img width="568" height="1084" alt="Image" src="https://github.com/user-attachments/assets/301fc9c1-69ff-470f-a488-ae6bb289e79a" />

컴퓨터와 인터랙션 중에도 상단부에는 제스쳐가 가능해서 조이스틱으로 이동이 가능한 버그가 발생합니다. 수정 가능할까요?

Gojaehyeon committed:
fix: 채팅 상황에서 조이스틱 생성되는 버그 수정

keon22han commented on NLP/NLP/Sources/Game/GameScene/GameScene.swift:
```diff
@@ -7,21 +7,33 @@ class GameScene: SKScene {
-            player.movePlayer(self.joyStick.getJoyStickMoveVector())
+            let moveVector = self.joyStick.getJoyStickMoveVector()
+            let strength = self.joyStick.getJoystickStrength()
+            let maxSpeed: CGFloat = 400 // 최대 속도
```
상수인 maxSpeed 는 Utils/Constants/ConstantValues.swift 에서 관리하도록 빼주어도 될 것 같습니다!

Gojaehyeon replied:
감사합니다.
수정완료!

keon22han commented on NLP/NLP/Sources/Game/GameScene/GameScene.swift:
```diff
@@ -7,21 +7,33 @@ class GameScene: SKScene {
+            let norm = sqrt(moveVector.x * moveVector.x + moveVector.y * moveVector.y)
+            let direction = norm > 0 ? CGVector(dx: moveVector.x / norm, dy: moveVector.y / norm) : .zero
+            let velocity = CGVector(dx: direction.dx * maxSpeed * strength, dy: direction.dy * maxSpeed * strength)
+            player.physicsBody?.velocity = velocity
```
update 메서드에 추가한 로직은 player.movePlayer() 메서드에서 수행하도록 수정해주세요. GameScene 에서는 플레이어 이동 로직이 보이지 않는게 좋을 것 같아용!

Gojaehyeon replied:
감사합니다.
수정완료!

keon22han commented on NLP/NLP/Sources/Game/GameScene/GameScene.swift:
```diff
@@ -7,21 +7,33 @@ class GameScene: SKScene {
+            if let gs = gameState, gs.isChatting { return }
+            self.joyStick.createDynamicJoystick(at: touchLocation, camera: camera)
+            isJoystickTouchActive = true
+            self.joyStick.startMove(touchLocation)
```
```swift
if self.joyStick.isJoyStickAvailableLocation(touchLocation) {
    isJoystickTouchActive = true
    self.joyStick.startMove(touchLocation)
    return
} 
if let gs = gameState, gs.isChatting { return }
self.joyStick.createDynamicJoystick(at: touchLocation, camera: camera)
isJoystickTouchActive = true
self.joyStick.startMove(touchLocation)
```
로 수정하는게 좋을 것 같습니다! 조기 리턴을 시키면
1. 가독성이 향상
2. 유지보수 용이
  - 조건별 흐름 명확 -> 추후 조건 추가 및 디버깅 시 헷갈리지 않음
3. 불필요한 코드 실행 방지
  - return 후 아래 코드 실행 방지 -> 로직 실수 줄임

등의 이점을 챙길 수 있습니당

Gojaehyeon replied:
감사합니다.
수정완료!

keon22han commented on NLP/NLP/Sources/Game/MainGameScene/MainGameScene.swift:
```diff
@@ -10,7 +10,6 @@ import SpriteKit
 class MainGameScene: GameScene {
     var box: BoxSprite?
     var computer: ChapOneComputerSprite?
-    weak var gameState: MainGameState?
```
요거 GameScene 을 상속받는 각 게임 Scene 마다 가져갈 상태(추후 ViewModel로 변경 예정)이라, GameScene에서 가지고 있으면 안됩니다!

Gojaehyeon replied:
감사합니다.
수정완료!

keon22han commented on NLP/NLP/Sources/Game/GameScene/GameScene.swift:
```diff
@@ -7,21 +7,33 @@ class GameScene: SKScene {
     var playerPhysicsType: PlayerPhysicsType = .normal
     var joyStick = JoyStick()
     var isJoystickTouchActive: Bool = false
+    weak var gameState: MainGameState?
```
요거 MainGameScene 으로 옮겨주셔야 합니다!

Gojaehyeon replied:
감사합니다.
수정완료!

keon22han commented on NLP/NLP/Sources/Game/Shapes/JoyStick.swift:
```diff
@@ -23,18 +25,53 @@ extension JoyStick {
+    
+    // MARK: - 조이스틱 생성 헬퍼 메서드
+    private func createJoystickAtPosition(_ position: CGPoint, camera: SKCameraNode) {
+        let radius: CGFloat = 100
```
요것도 ConstantValues.swift 파일에서 관리하면 좋겠네용

Gojaehyeon replied:
감사합니다.
수정완료!

keon22han reviewed (commented):
수고하셨습니다 고! 코멘트 확인해주세요!!

Gojaehyeon committed:
refactor: maxspeed 상수로 관리하도록 수정

Gojaehyeon committed:
refactor: 플레이어 이동 로직 PlayerSprite로 위임

- GameScene의 update에서 직접 플레이어 속도를 계산하던 로직을 제거
- 이동 방향, 세기, 최대 속도 계산 후 player.movePlayer(direction:strength:maxSpeed:)로 위임
- PlayerSprite 내부에서 실제 물리 velocity를 설정하도록 구조 개선
- GameScene은 이동 명령만 내리고, 이동 구현은 PlayerSprite가 담당하도록 역할 분리

Gojaehyeon committed:
refactor: 조기 리턴 도입으로 조이스틱 생성 로직 개선

- joyStick 위치 유효성 판단 후 조기 리턴 적용
- gameState.isChatting 조건도 조기 리턴 처리
- 조건 분기를 명확히 하여 추후 조건 추가 및 디버깅 용이
- 불필요한 코드 실행 방지로 로직 실수 가능성 줄임

Gojaehyeon committed:
refactor: GameScene/ MainGameScene - gameState 관리 구조 개선

- gameState(weak var gameState: MainGameState?) 선언을 GameScene에서 제거하고 MainGameScene에서만 관리하도록 구조 변경
- GameScene에서는 gameState 관련 코드 및 대화 상태 체크 로직 완전히 제거
- MainGameScene에서만 gameState를 선언하고, touchesBegan을 override하여 대화(채팅) 중에는 동적 조이스틱이 생성되지 않도록 구현
- 각 게임 Scene이 독립적으로 상태를 관리할 수 있도록 구조 개선
- 코드 명확성 및 유지보수성 향상

Gojaehyeon committed:
refactor: 조이스틱 생성 헬퍼 메서드 이동

keon22han reviewed (approved):
Good boy
